### PR TITLE
add route table option in static route for subnet

### LINKED ISSUE #2637: multiple route tables support
### Feature request

Shell we have multiple route tables in VPC level and can be associate/disassociate with Subnet.
I think this features is already in OVN (set static route with --route_table in lr and set option in lrp)

### Use case

- reuse routings and use different route table in each Subnet
- associate / disassociate routing table in Subnet easily

I think use policy-based route with src=<subnet_cidr> match can instead, but will be more flow entry and difficulty to switch route table

I can contribute to this feature If that make sense

### PR DESCRIPTION
**Pull request by mingoooo**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

### What type of this PR
Examples of user facing changes:
- Features

### Which issue(s) this PR fixes:
Fixes #2637

### <samp>🤖 Generated by Copilot at 9e1ac75</samp>

This pull request adds support for multiple route tables per subnet and VPC in the `kube-ovn` CRDs, controller, and OVN database. It modifies the existing functions and commands that deal with static routes to accept and use the route table name as a parameter. It also adds new functions to get and set the route table name of router ports and to reconcile the route table name for subnets. It updates the garbage collection and migration logic to handle multiple route tables. It also updates the test file to use the `util` package and check the route table name.

### <samp>🤖 Generated by Copilot at 9e1ac75</samp>

> _The pull request is quite a big deal_
> _It adds route tables for NAT and subnet appeal_
> _With `controller` and `ovs` changes_
> _And `v1` structs rearranges_
> _It makes static routes more flexible and real_

### <samp>🤖 Generated by Copilot at 9e1ac75</samp>

*  Add a new field `routeTable` to the `VpcNatGateway` and `Subnet` CRD schemas to allow users to specify a custom route table for the NAT gateway and the subnet ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-d671263a0aa28dac78faa045e43d18cde05bf81846c6d1bf1db009cddda883f2R1233-R1234), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-d671263a0aa28dac78faa045e43d18cde05bf81846c6d1bf1db009cddda883f2R1714-R1715))
* Add a new field `RouteTable` to the `SubnetStatus` and `StaticRoute` structs in the `v1` package to store the route table name for the subnet status and the static route spec ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aR164-R165), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aL396-R403))
* Add a new parameter `routeTable` to the `checkRouteExist`, `deleteStaticRoute` and `DeleteStaticRoute` functions in the `controller` package to pass the route table name to the functions that check, delete or add a static route in the OVN database ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-8a92aa9e5847a415c4c90271fa845087d27a67fd272b8efc2636b06d52cb3619L832-R832), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L806-R807), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1723-R1750))
* Add a new argument `routeTable` to the `AddStaticRoute`, `DeleteStaticRoute`, `DeleteMatchedStaticRoute` and `DeleteStaticRouteByNextHop` functions in the `ovs` package to pass the route table name to the OVN commands that add or delete a static route in the OVN database ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373R154-R157), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373L179-R189), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373L202-R220), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373L625-R684), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373L648-R696), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373L666-R714))
* Add a new argument `routeTable` to the `AddStaticRoute`, `DeleteStaticRoute`, `DeleteMatchedStaticRoute` and `checkRouteExist` function calls in the `controller` package to pass the route table name to the functions that add or delete a static route in the OVN database ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-8a92aa9e5847a415c4c90271fa845087d27a67fd272b8efc2636b06d52cb3619L1037-R1037), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-8a92aa9e5847a415c4c90271fa845087d27a67fd272b8efc2636b06d52cb3619L1045-R1047), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-449e8d1bd46b9c978e208dde912a1d7a76d17ce553067f8d0a67a27f17338d26L741-R743), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-449e8d1bd46b9c978e208dde912a1d7a76d17ce553067f8d0a67a27f17338d26L782-R792), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-449e8d1bd46b9c978e208dde912a1d7a76d17ce553067f8d0a67a27f17338d26L871-R876), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1342-R1359), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1349-R1366), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1467-R1489), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1476-R1502), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL284-R302), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL323-R337), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecR343), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL343-R354), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL352-R363), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL358-R369))
* Add a new condition `route.RouteTable == item.RouteTable` or `route.RouteTable == routeTable` to the `gcStaticRoute`, `reconcileVpcUseBfdStaticRoute`, `reconcileVpcAddNormalStaticRoute` and `checkRouteExist` functions in the `controller` package to check if the route table name matches when comparing the static routes in the OVN database, the VPC spec or the default VPC ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-efdb6355f2187b31e5e61610c24bf4d446a23d814d51448c170a91ef96ae1e89L631-R631), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1165-R1172), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1222-R1275), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-8a92aa9e5847a415c4c90271fa845087d27a67fd272b8efc2636b06d52cb3619L844-R844))
* Add a new condition `oldSubnet.Spec.RouteTable != newSubnet.Spec.RouteTable` to the `enqueueUpdateSubnet` function in the `controller` package to check if the route table name of the subnet has changed and enqueue the subnet for update if so ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L102-R103))
* Add a new condition `item.RouteTable == ""` to the `formatVpc` function in the `controller` package to check if the route table name of the static route is empty and set it to the main route table name if so ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL566-R580))
* Add a new field `RouteTable: subnet.Spec.RouteTable` or `RouteTable: routeTable` to the `StaticRoute` struct initialization in the `controller` and `ovs` packages to set the route table name for the static route that will be added to the VPC spec or parsed from the OVN command output ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1173-R1185), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL537-R553), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373L378-R404), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373L407-R442))
* Add a new block of code to the `migrateNodeRoute` and `delLearnedRoute` functions in the `controller` package to list all the route tables of the cluster router or the logical router and delete the obsolete or learned static route from each route table, instead of only deleting from the main route table ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-92aa62435bc2b1b2be34f478f650597068bcf35b8d5975d119f91cf78ac54988L763-R775), [link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-173d2c52eb78053b84437284c077aba7452636bc799728cc475cc343ea70c2ceL382-R394))
* Add a new block of code to the `AddStaticRoute` function in the `ovs` package to check if the route table name is the main route table name and set it to an empty string if so, because the OVN database uses an empty string for the main route table name ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373R154-R157))
* Add a new block of code to the `parseLrRouteListOutput` function in the `ovs` package to check if the line matches the route table name pattern and set the route table name variable accordingly, or use the main route table name as the default value ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373R410-R419))
* Add a new function `reconcileRouteTableForSubnet` to the `controller` package to define a function that reconciles the route table name for the subnet and updates the router port options accordingly ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205R2531-R2563))
* Add a new function call `c.reconcileRouteTableForSubnet(subnet)` to the `reconcileSubnet` function in the `controller` package to call the function that reconciles the route table name for the subnet and updates the router port options accordingly ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205R944-R948))
* Add a new function `getRouteTablesByVpc` to the `controller` package to define a function that returns a map of route table names and static routes for a given VPC ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecR587-R590))
* Add a new function `getStaticRouteItemKey` to the `controller` package to define a function that returns a unique key for a static route item based on its policy, CIDR, next hop and route table ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecR962-R973))
* Add a new function `GetRouteTables` to the `ovs` package to define a function that interacts with the OVN database to get the route tables of a router ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373R2060-R2102))
* Add a new function `SetRouterPortOptions` to the `ovs` package to define a function that interacts with the OVN database to set the options of a router port ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373R2060-R2102))
* Add a new function `UpdateRouterPortRouteTable` to the `ovs` package to define a function that interacts with the OVN database to update the route table name of a router port ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373R2060-R2102))
* Add a new constant `mainRouteTableName` to the `ovs` package to define a constant for the main route table name in the OVN database ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-a90d8643bea46e8b41ae2967284c3fbc3f11681b2ef2ae06d7406815f0a20373R29-R30))
* Add a new import `github.com/kubeovn/kube-ovn/pkg/util` to the `ovs` package test file to use the `util` package constants and functions in the test cases ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-b874caae9f53d51183c84dcfd89891600d1e298c48947db8dfe4104db72c28abR4))
* Add a new field `RouteTable: util.MainRouteTable` to the `StaticRoute` struct initialization in the `Test_parseLrRouteListOutput` function in the `ovs` package test file to set the expected route table name for the static route in the test case ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-b874caae9f53d51183c84dcfd89891600d1e298c48947db8dfe4104db72c28abL11-R54))
* Add a new assertion `ast.Equal(routeList[0].RouteTable, util.MainRouteTable)` to the `Test_parseLrRouteListOutput` function in the `ovs` package test file to check if the parsed route table name matches the expected value in the test case ([link](https://github.com/kubeovn/kube-ovn/pull/2748/files?diff=unified&w=0#diff-b874caae9f53d51183c84dcfd89891600d1e298c48947db8dfe4104db72c28abR65))

### CODE REVIEW

File: pkg/controller/gc.go [hunk 1]
```diff
@@ -642,8 +642,8 @@ func (c *Controller) gcStaticRoute() error {
 				klog.Errorf("failed to get NatRule by LogicalIP %s, %v", route.CIDR, err)
 				continue
 			}
-			klog.Infof("gc static route %s %s %s", route.Policy, route.CIDR, route.NextHop)
-			if err := c.ovnLegacyClient.DeleteStaticRoute(route.CIDR, c.config.ClusterRouter); err != nil {
+			klog.Infof("gc static route %s %s %s %s", route.Policy, route.CIDR, route.NextHop, route.RouteTable)
```
Comment (bobz965) [inline pkg/controller/gc.go:645]:
```suggestion
			klog.Infof("gc static route %s %s %s %s", route.RouteTable, route.Policy, route.CIDR, route.NextHop)
```

File: pkg/ovs/ovn-nbctl-legacy.go [hunk 2]
```diff
@@ -622,21 +655,33 @@ func (c *LegacyClient) SnatRuleExists(eip, ipCidr string) (bool, error) {
 	return false, fmt.Errorf("snat rule not exist")
 }
 
-func (c LegacyClient) DeleteMatchedStaticRoute(cidr, nexthop, router string) error {
+func (c LegacyClient) DeleteMatchedStaticRoute(cidr, nexthop, router, routeTable string) error {
+	if routeTable == util.MainRouteTable {
```
Comment (bobz965) [inline pkg/ovs/ovn-nbctl-legacy.go:659]: if MainRouteTable is "" in the ovn nb, how about set MainRouteTable = "" in the const.go

Comment (mingoooo) [reply pkg/ovs/ovn-nbctl-legacy.go]: > if MainRouteTable is "" in the ovn nb, how about set MainRouteTable = "" in the const.go

good idea, already changed

**Review by bobz965 — APPROVED** [review]

**Comment by oilbeater** [comment]
@mingoooo Great work! A document about how to use it should also be added [to the vpc guide.](https://kubeovn.github.io/docs/v1.12.x/guide/vpc/)

**Comment by mingoooo** [comment]
> @mingoooo Great work! A document about how to use it should also be added [to the vpc guide.](https://kubeovn.github.io/docs/v1.12.x/guide/vpc/)

ok, created a pr please review - https://github.com/kubeovn/docs/pull/77

**Comment by oilbeater** [comment]
@mingoooo the conflict should be resolved

**Comment by mingoooo** [comment]
> @mingoooo the conflict should be resolved

done

**Pull request merged**